doc/man/id3tag.1
.TH ID3TAG 1 "May 2000" local "User Command"
.SH NAME
id3tag \- Tags an mp3 file with id3v1 and/or id3v2 tags.
.SH SYNOPSIS
.B id3tag
.RB [
.I OPTION
.RB ] ...
.RB [
.I FILE
.RB ] ...
.br
.SH DESCRIPTION
.B Id3tag
will render both types of tag by default.  Only the last
tag type indicated in the option list will be used.  Non-
rendered will remain unchanged in the original file.  Will
also parse and convert Lyrics3 v2.0 frames, but will not
render them.

.SH OPTIONS
.TP
.B \-1, \-\-v1tag
Render only the id3v1 tag
.TP
.B \-2, \-\-v2tag
Render only the id3v2 tag
.TP
.B \-h, \-\-help
Display help and exit
.TP
.B \-v, \-\-version
Display version information and exit
.TP
.B \-a, \-\-artist ARTIST
Set the artist information
.TP
.B \-s, \-\-song SONG
Set the song title information
.TP
.B \-A, \-\-album ALBUM
Set the album title information
.TP
.B \-c, \-\-comment COMMENT
Set the comment information
.TP
.B \-C, \-\-desc DESCRIPTION
Set the comment description
.TP
.B \-g, \-\-genre num
Set the genre number
.TP
.B \-y, \-\-year num
Set the year
.TP
.B \-t, \-\-track num
Set the track number
.TP
.B \-T, \-\-total num
Set the total number of tracks on the album

.SH SEE ALSO
id3convert(1), id3info(1), id3v2(1)
.SH AUTHOR
.B id3lib
was originally designed and implemented by Dirk Mahoney and is
maintained by Scott Thomas Haug <sth2@cs.wustl.edu>. Manual page written for
Debian GNU/Linux by Robert Woodcock <rcw@debian.org>.



ChangeLog
2006-02-17  Jerome Couderc

    * Patch from Spoon to fix UTF-16 writing bug
      http://sourceforge.net/tracker/index.php?func=detail&aid=1016290&group_id=979&atid=300979

2003-03-02 Sunday 17:38   Thijmen Klok <thijmen@id3lib.org>

	* THANKS (1.20): added more people 

2003-03-02 Sunday 17:37   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32 (1.25), config.h.win32.in (1.13), configure
	(1.79), configure.in (1.68), macconfig.h (1.12): changes to reflect
	new version 

2003-03-02 Sunday 17:19   Thijmen Klok <thijmen@id3lib.org>

	* config.sub (1.3), config.guess (1.3): update from
	ftp://ftp.gnu.org/pub/gnu/config/ timestamp 2003-02-22 

2003-03-02 Sunday 16:17   Thijmen Klok <thijmen@id3lib.org>

	* examples/: demo_main.cpp (1.4), test_io.cpp (1.6),
	test_remove.cpp (1.13): added cerr cout and endl where needed 

2003-03-02 Sunday 15:39   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_find.cpp (1.28): Fixes a bug where a string is tried to
	be made from NULL 

2003-03-02 Sunday 15:30   Thijmen Klok <thijmen@id3lib.org>

	* src/header_tag.cpp (1.25): adds a check on _info for NULL 

2003-03-02 Sunday 15:23   Thijmen Klok <thijmen@id3lib.org>

	* src/: field_string_ascii.cpp (1.29), field_string_unicode.cpp
	(1.33), field_binary.cpp (1.27): added a check for NULL in Set() 

2003-03-02 Sunday 15:14   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_file.cpp (1.43): fixes left open file 

2003-03-02 Sunday 14:56   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/id3lib_strings.h (1.4): fix for undefined wchar_t
	template 

2003-03-02 Sunday 14:35   Thijmen Klok <thijmen@id3lib.org>

	* src/tag.cpp (1.55), include/id3/tag.h (1.63): fix because
	GetFileName need to return a pointer which keeps to be valid 

2003-02-21 Friday 04:47  slackorama

	* include/id3/globals.h (1.54): added a missing const to remove
	warning when compiling with -Wwrite-strings (as is done in KDE and
	many other projects) 

2002-11-24 Sunday 20:39   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.138): added all entries for 3.8.2 release 

2002-11-24 Sunday 20:34   Thijmen Klok <thijmen@id3lib.org>

	* id3com/id3com.dsp (1.20): added link32 to zlib 

2002-11-24 Sunday 18:33   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse.cpp (1.47): fixed a bug which caused v1 tags to be
	ignored after a lyrics v2 tag 

2002-11-24 Sunday 17:31   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse.cpp (1.46): fixed a stupid bug of mine, affecting
	wrong handling of id3v1 tags 

2002-11-03 Sunday 01:41   Thijmen Klok <thijmen@id3lib.org>

	* src/field.cpp (1.47): fixes access to invalid framedefs 

2002-11-02 Saturday 23:28   Thijmen Klok <thijmen@id3lib.org>

	* id3com/win32.readme.first.txt (1.4),
	libprj/win32.readme.first.txt (1.4), prj/win32.readme.first.txt
	(1.4): improved comment on delphi 

2002-11-02 Saturday 22:02   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_file.cpp (1.42): fixed a bug in RenderV2ToFile, file was
	not cleared and thus eof mark was still set 

2002-11-02 Saturday 19:10   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.137): added todays work 

2002-11-02 Saturday 19:07   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32 (1.24), configure (1.78), configure.in (1.67),
	macconfig.h (1.11): bumped version up to 3.8.1 

2002-11-02 Saturday 19:03   Thijmen Klok <thijmen@id3lib.org>

	* id3lib.spec.in (1.27): added a line on mp3 headers 

2002-11-02 Saturday 18:49   Thijmen Klok <thijmen@id3lib.org>

	* THANKS (1.19): added latest help 

2002-11-02 Saturday 18:48   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/globals.h (1.53), src/mp3_header.h (1.4),
	src/mp3_parse.cpp (1.6): added vbr support 

2002-11-02 Saturday 18:35   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/tag.h (1.62), src/tag.cpp (1.54), src/tag_file.cpp
	(1.41), src/tag_impl.h (1.10), src/tag_parse.cpp (1.45): added
	streaming support 

2002-11-02 Saturday 18:03   Thijmen Klok <thijmen@id3lib.org>

	* libprj/win32.readme.first.txt (1.3),
	id3com/win32.readme.first.txt (1.3), prj/win32.readme.first.txt
	(1.3): added comment about id3lib.pas 

2002-11-02 Saturday 17:56   Thijmen Klok <thijmen@id3lib.org>

	* delphi/Id3lib.pas (1.1): added delphi interface to id3lib 

2002-11-02 Saturday 13:45   Thijmen Klok <thijmen@id3lib.org>

	* id3com/: ID3Tag.cpp (1.14), id3com.dsp (1.19), id3com.idl (1.13): 
	Updated id3com in line with latest id3lib 

2002-10-04 Friday 10:57   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.136): added latest entries 

2002-10-04 Friday 10:52   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_parse_lyrics3.cpp (1.35): cleaning source 

2002-09-23 Monday 23:32  slackorama

	* src/tag_parse_lyrics3.cpp (1.34): Fixed a "version priority"
	issue in the ID3 lyrics tagging stuff.  Posted patch on id3lib-devel
	first.  

2002-09-21 Saturday 19:23   Thijmen Klok <thijmen@id3lib.org>

	* include/id3.h (1.21), include/id3/tag.h (1.61), src/c_wrapper.cpp
	(1.22, rel-3-7-10), src/helpers.cpp (1.12), src/tag.cpp (1.53),
	src/tag_impl.cpp (1.13), src/tag_impl.h (1.9): added boolean return
	for AttachFrame 

2002-09-21 Saturday 18:04   Thijmen Klok <thijmen@id3lib.org>

	* src/tag_impl.cpp (1.12): initializing _mp3_info to NULL 

2002-09-19 Thursday 12:20   Thijmen Klok <thijmen@id3lib.org>

	* src/misc_support.cpp (1.39), include/id3/misc_support.h (1.29): 
	fixed prototpyes ID3_GetSyncLyrics 

2002-08-13 Tuesday 18:10   Thijmen Klok <thijmen@id3lib.org>

	* include/id3/id3lib_strings.h (1.3): fixed compile for gcc 3.2 

2002-08-11 Sunday 12:58   Thijmen Klok <thijmen@id3lib.org>

	* cvs2cl.pl (1.2): changed command for use of ssh 

2002-08-10 Saturday 22:39   Thijmen Klok <thijmen@id3lib.org>

	* config.h.win32.in (1.12): updated for version info on dll 

2002-08-10 Saturday 22:13   Thijmen Klok <thijmen@id3lib.org>

	* ChangeLog (1.135): updated changelog for latest changes before
	final 

2002-08-10 Saturday 22:02   Thijmen Klok <thijmen@id3lib.org>


examples/id3tag.cpp
// SPDX-License-Identifier: LGPL-3.0-only

#include <config.h>

#include <id3/id3lib_streams.h>
#include <id3/misc_support.h>
#include <id3/tag.h>
#include <map>
#include <stdlib.h>
#include <string.h>
#include <vector>

#define APP_NAME "id3tag"

using std::cout;
using std::endl;
using std::string;

/* Replaced gengetopt as with demo_info/id3info */
void PrintUsage ()
{
    cout << "Will render both types of tag by default.  Only the last" << endl;
    cout << "tag type indicated in the option list will be used.  Non-" << endl;
    cout << "rendered will remain unchanged in the original file.  Will" << endl;
    cout << "also parse and convert Lyrics3 v2.0 frames, but will not" << endl;
    cout << "render them." << endl;
    cout << endl;
    cout << "Usage: " << APP_NAME << " [OPTIONS]... <FILE>..." << endl;
    cout << "       " << APP_NAME << " [-h | -V]" << endl;
    cout << endl;
    cout << "  -aSTRING   --artist=STRING   Set the artist information" << endl;
    cout << "  -ASTRING   --album=STRING    Set the album title information" << endl;
    cout << "  -sSTRING   --song=STRING     Set the title information" << endl;
    cout << "  -cSTRING   --comment=STRING  Set the comment information" << endl;
    cout << "  -CSTRING   --desc=STRING     Set the comment description" << endl;
    cout << "  -ySTRING   --year=STRING     Set the year" << endl;
    cout << "  -tSTRING   --track=STRING    Set the track number" << endl;
    cout << "  -TSTRING   --total=STRING    Set the total number of tracks" << endl;
    cout << "  -gSHORT    --genre=SHORT     Set the genre" << endl;
    cout << "  -1  --v1tag     Render only the id3v1 tag (default=off)" << endl;
    cout << "  -2  --v2tag     Render only the id3v2 tag (default=off)" << endl;
    cout << endl;
    cout << "  -h, --help      Print help and exit" << endl;
    cout << "  -V, --version   Print version and exit" << endl;
    cout << endl;
}

void PrintVersion ()
{
    cout << APP_NAME << " version: " << VERSION << endl;
    cout << "Tags mp3 files with ID3v1/1.1 and/or id3v2 tags." << endl;
    cout << "Uses " << ID3LIB_FULL_NAME << endl;
    cout << endl;
}

void DisplayTags (ostream &os, luint nTags)
{
    if (! ( (nTags & ID3TT_ID3V1) || (nTags & ID3TT_ID3V2))) {
        os << "no tag";
    }

    if (nTags & ID3TT_ID3V1) {
        os << "v1";
    }

    if ( (nTags & ID3TT_ID3V1) && (nTags & ID3TT_ID3V2)) {
        os << " and ";
    }

    if (nTags & ID3TT_ID3V2) {
        os << "v2";
    }
}

int main (int argc, char *const argv[])
{
    // int ulFlag = ID3TT_ID3;

    if (argc < 2) {
        PrintUsage();
        exit (1);
    }

    int ulFlag = ID3TT_ID3;

    std::map<string, string> argmap;
    argmap["a"] = "ARTIST";
    argmap["artist"] = "ARTIST";
    argmap["A"] = "ALBUM";
    argmap["album"] = "ALBUM";
    argmap["s"] = "SONG";
    argmap["song"] = "SONG";
    argmap["c"] = "COMMENT";
    argmap["comment"] = "COMMENT";
    argmap["C"] = "DESCRIPTION";
    argmap["desc"] = "DESCRIPTION";
    argmap["y"] = "YEAR";
    argmap["year"] = "YEAR";
    argmap["t"] = "TRACK";
    argmap["track"] = "TRACK";
    argmap["T"] = "TOTAL";
    argmap["total"] = "TOTAL";
    argmap["g"] = "GENRE";
    argmap["genre"] = "GENRE";


    std::vector<string> files;
    std::map<string, string> tagmap;
    string key, value, tagkey;
    string::size_type sep;

    for (size_t i = 1; i < argc; i++) {
        std::string arg = argv[i];

        if (arg == "-h" || arg == "--help") {
            PrintUsage();
            exit (0);
        } else if (arg == "-V" || arg == "--version") {
            PrintVersion();
            exit (0);
        } else if (arg == "-1" || arg == "--v1tag") {
            ulFlag = ID3TT_ID3V1;
        } else if (arg == "-2" || arg == "--v2tag") {
            ulFlag = ID3TT_ID3V2;
            // process long options first --artist=X
        } else if (arg[0] == '-' && arg[1] == '-') {
            sep = arg.find ("=");

            if (sep < 3)
                continue;

            key = arg.substr (2, sep - 2);
            value = arg.substr (sep + 1);
            tagkey = argmap[key];

            if (tagkey == key) {
                cout << "Invalid key: " << key << endl;
                continue;
            }

            tagmap[tagkey] = value;
            // process short options -aARTIST_NAME
        } else if (arg[0] == '-') {
            key = arg[1];
            value = arg.substr (2);
            tagkey = argmap[key];

            if (tagkey == key) {
                cout << "Invalid key: " << key << endl;
                continue;
            }

            tagmap[tagkey] = value;
        } else {
            files.push_back (arg);
        }
    }

    std::map<string, string>::iterator mapit;
    std::vector<std::string>::iterator i;

    for (i = files.begin(); i != files.end(); i++) {
        std::string filename = *i;
        ID3_Tag myTag;

        cout << "Tagging " << filename << ": ";

        myTag.Link (filename.c_str());

        cout << "attempting ";
        DisplayTags (cout, ulFlag);
        cout << endl;

        mapit = tagmap.find ("ARTIST");

        if (mapit != tagmap.end()) {
            ID3_AddArtist (&myTag, (*mapit).second.c_str(), true);
            cout << "+++ Artist  = " << (*mapit).second << endl;
        }

        mapit = tagmap.find ("ALBUM");

        if (mapit != tagmap.end()) {
            ID3_AddAlbum (&myTag, (*mapit).second.c_str(), true);
            cout << "+++ Album  = " << (*mapit).second << endl;
        }

        mapit = tagmap.find ("SONG");

        if (mapit != tagmap.end()) {
            ID3_AddTitle (&myTag, (*mapit).second.c_str(), true);
            cout << "+++ Song  = " << (*mapit).second << endl;
        }

        mapit = tagmap.find ("YEAR");

        if (mapit != tagmap.end()) {
            ID3_AddYear (&myTag, (*mapit).second.c_str(), true);
            cout << "+++ Year  = " << (*mapit).second << endl;
        }

        mapit = tagmap.find ("COMMENT");

        if (mapit != tagmap.end()) {
            value = (*mapit).second;
            string desc;
            mapit = tagmap.find ("DESCRIPTION");

            if (mapit != tagmap.end())
                desc = (*mapit).second;

            ID3_AddComment (&myTag, value.c_str(), desc.c_str(), true);
            cout << "+++ Comment = " << value << endl;

            if (desc.length()) {
                cout << "+++ Comment Description" << endl;
                cout << "            = " << desc << endl;
            }
        }

        mapit = tagmap.find ("GENRE");

        if (mapit != tagmap.end()) {
            unsigned short genre = ::strtol ( (*mapit).second.c_str(), NULL, 10);
            ID3_AddGenre (&myTag, genre, true);
            cout << "+++ Genre  = " << genre << endl;
        }

        mapit = tagmap.find ("TRACK");

        if (mapit != tagmap.end()) {
            unsigned short track = ::strtol ( (*mapit).second.c_str(), NULL, 10);
            string totalstr;
            mapit = tagmap.find ("TOTAL");

            if (mapit != tagmap.end())
                totalstr = (*mapit).second;

            unsigned short total = ::strtol (totalstr.c_str(), NULL, 10);
            ID3_AddTrack (&myTag, track, total, true);
            cout << "+++ Track = " << track << endl;
        }

        luint nTags = myTag.Update (ulFlag);
        cout << ", tagged ";

        DisplayTags (cout, nTags);

        cout << endl;
    }

    return 0;
}
